### ISSUE TITLE
Textpattern 4.8.2 release flight plan

### ISSUE THREAD

**Issue by petecooper** [textpattern/textpattern#1513]
We are targeting Textpattern 4.8.2 for release in Q3 2020.

Please update this post and subsequent replies with tasks etc as needed, assigning yourself or others where appropriate.

## Preflight

**Repo Issues**

* [`textpattern/textpattern`](https://github.com/textpattern/textpattern/milestone/14)
* `textpattern/textpacks` (n/a)
* [`textpattern/textpattern.github.io`](https://github.com/textpattern/textpattern.github.io/milestone/2)

- [x] latest pophelp (@philwareham)
- [x] write .com blog post (@petecooper and @Bloke - ID 430, awaiting sanity check and additional late changes from `HISTORY.txt`) 

## Take Off

- [x] package up for GitHub and .com (@Bloke)
- [x] update demo site (@petecooper)
- [x] publish .com blog post: ID 430 (anyone)
- [x] forum announcement (@petecooper)
- [x] forum feedback (@petecooper)
- [x] socials (@petecooper)
- [x] tag public/admin themes with release tag (@philwareham)
- [x] tag Textpacks and Pophelps with release tag (@philwareham)

**Comment by philwareham** [comment]
Anyone got a date they’d like to aim for? Just do I can make sure my bits are ticked off in time? Ta.

**Comment by petecooper** [comment]
I'll make a start on the blog post over the weekend.

**Comment by petecooper** [comment]
Given the slightly precarious nature of the repo, and not wanting to make work unnecessarily, I propose we bring forward 4.8.2 release to sooner rather than later, jettison 4.8.2 branch, and start 4.8.3 as a maintenance release. Reasonable?

**Comment by Bloke** [comment]
Definitely. I wanna get 4.8.2 out this weekend if we can. @bloatware @philwareham Any objections?

You've got mail, btw, Pete.

**Comment by bloatware** [comment]
No objection, but a stress-test appeal, to avoid 4.8.3.

**Comment by petecooper** [comment]
>a stress-test appeal, to avoid 4.8.3.

This sounds eminently reasonable. You have much better visibility than I do, @bloatware – is 4.8 'ready' after 4.8.2?

**Comment by Bloke** [comment]
In which case, if we're comfy with the `gTxt()` hacks in txp_auth for the time being, let's commit-freeze 4.8.2 now. No tweaks, no performance improvements or features unless they're directly related to an open issue or a bug report.

Given the state of the 4.8.2 branch I think the only solution is to rebuild it, replaying relevant commits from 'this is 4.8.1' forward in a fresh branch based off `master`. And then hope/pray the new branch merges with dev at the end of the process.

We can ignore the 'merge 4.8.2 into 4.82' type of commits, as they add no value. But I've spotted at least one commit that shouldn't be in 4.8.2 (8f2bf1e0bc). Only a fraction of that commit should be there. No idea how that even got in there. I think it's part of the botched merges and reverts that have gone in today (or some glitch in the history, as Microsoft tinker behind the scenes with GitHub). That commit can't possibly exist at that point, since it makes changes to lines of code that aren't even in 4.8.2!

There may be other commits like that, so whoever goes through replaying commits to a new branch needs to be mindful of which ones are actually relevant to 4.8.2.

Once the new branch is made, I'll launch a flare and I suggest at that point we all take a flamethrower to our local 4.8.2 branches.

Agreed?

**Comment by bloatware** [comment]
@petecooper I hope so, playing with two branches is not my cup of tea. I guess it will depend (modulo bugs) on how disruptive/timely 4.9 will be. But my crystal ball is hardly bigger than yours.

@Bloke sorry, I don't quite remember what the problem with `txp_auth.php` was. Commenting out two `gTxt('nothing')` calls does not seem to change anything.

**Comment by Bloke** [comment]
Ack, I just went back to 'this is 4.8.1' on GitHub and scrolled up a bit on the 4.8.2 branch, opening some commits at random. There are loads of 4.9-related commits in the 4.8.2 branch. What the hell...?!?!

**Comment by Bloke** [comment]
> Commenting out two `gTxt('nothing')` calls does not seem to change anything.

Huh, now we've switched to `swapStrings()` it seems you're right. I must have inadvertently fixed this compared to when it first appeared. Sorry for causing you trouble.

I'll make sure when we trawl the commits that I remove those two hacky lines in a new commit at the end of the process. Woohoo.

**Comment by Bloke** [comment]
@bloatware @philwareham @petecooper 

Can we get away with 351439b44 in 4.8.2? I like the multi-edit widget not being available until you select some checkboxes. It was previously only in 4.9 branch. Is there anything stopping us bringing it into 4.8.2?

**Comment by philwareham** [comment]
Nothing stopping you as far as my work - don’t think it has any special new css.

**Comment by petecooper** [comment]
Release article is drafted and ready for more eyes.

**Comment by bloatware** [comment]
@Bloke I don't see why https://github.com/textpattern/textpattern/commit/351439b4486a28697f338e6b62daf7813f52d2ba couldn't go into 4.8.2, if GitHub has no objections.

**Comment by Bloke** [comment]
Thanks. Looks like that commit works okay in 4.8.2.

All we gotta do now is figure out how to fix the repos. Investigation ongoing. Please everyone refrain from any commit activity for the foreseeable until I've managed to recreate things in a stable manner. Thanks!

**Comment by Bloke** [comment]
Well that was a fun day. tl;dr:

* `4.8.2` branch is borked. **DO NOT USE IT** Once we are sure it has nothing of value that can be copied (NOT cherry picked) from it, I will delete it from GitHub and I urge you to delete your own local branches.
* In the next few days before we release 4.8.2, use the `482-recovery-bloke` branch instead as our de facto 4.8.2 branch. That's a clean-built branch from master up to today.

@philwareham Please would you check in the `482-recovery-bloke` branch that the admin-theme CSS files and the setup-multisite.css are correct as per the 4.8.2 admin theme repos, as I had to manually merge those files. Thank you, and sorry for any hassle caused.

@petecooper Please would you rebuild the 482 demo using `482-recovery-bloke` as the build point. Thank you.

**Everyone** Please checkout the `482-recovery-bloke` branch and test it to destruction. Installs, upgrades, features, functionality, front-end tags, admin side, everything. It's feature frozen, so bug fixes only please. And don't worry about syncing it to dev for now. Although it should work fine now that I've gone through the pain, let's not tempt fate :) I'll mop up any latent commits into dev when we do the final stable release.

Longer story, for those interested:

Pete and I tried many different ways to resuscitate the 4.8.2 branch. One thing we thought would work was to effectively roll back the changes of the day and start afresh like this:

1) git clone the borked repo NOW.
2) Download the dev files from GitHub at the last known-good commit point yesterday evening.
3) Download the 4.8.2 files from GitHub at the last known-good commit point yesterday evening.
4) Copy the 4.8.2 files from last known good over the borked 4.8.2 branch and commit: thus creating a fresh start for 4.8.2.
5) Copy the dev files from last known good over the borked dev branch and commit: thus creating a fresh start for dev.
6) Replay the commits of today into the 4.8.2 branch by hand.
7) Merge 4.8.2 to dev.

Did it work? No. Two reasons:

a) GitHub don't seem to allow you to download a branch-specific version of the state of a repo (or if they do, it's a well hidden feature). Luckily, Pete found that Tower for Mac could do this, so he zipped up a known-good dev and known-good 4.8.2 for me. Yay.
b) After using `rsync -rv --size-only` to copy just the changed files from these known-good archives, over first 4.8.2 and then dev branches and committing, I expected to then be able to do a regular `git merge 4.8.2` into dev. But it didn't work. It only merged a single file (and had a bunch of conflicts at that).

The reason for the latter behaviour, I suspect, is that a merge goes back and looks at the history of the branches you're merging to find out at what point they diverged. The last common divergent point of 4.8.2 and dev was _before_ the rollback/fresh start I just did.

So instead of the merge going back a couple of commits and winding forward from the rollback point, it went back to what I'm calling the borkfork point - the time when dev was accidentally partially merged into 4.8.2 - and ran the merge from there. Since most of the changes were already in both branches in various guises due to the rollbacks and commit/reverts of the day, it only detected a single file as having changed.

Faced with the prospect of either:

1) Making a fresh branch from master 4.8.1 up to today by replaying a series of specific commits into it; or
2) Making a fresh branch from dev and reverting every 4.9-related commit all the way back to master 4.8.1, thus leaving only 4.8.2 commits intact

I chose the first option, built that, and went through the initial merge conflict pain of syncing that back into dev. Hence, 482-recovery-bloke is now our new 4.8.2 branch.

Lesson learned. When git works, it's fantastic. When a slip-up happens, it punches you in the nads and doesn't stop.

**Comment by philwareham** [comment]
Sounds hideous - thanks for slogging through a fix. So it was a git cherry-pick of one simple text file that caused all this?!

**Comment by Bloke** [comment]
Dunno what caused it. Might have been the cherry-pick. Might have been a misclick in the merge to cause it to merge in the wrong direction. Might have been Microsoft. Might have been git just having enough of our back and forths and getting its knickers in a twist.

At least it was fixable... eventually. Sorry it took so long.

**Comment by petecooper** [comment]
I've been back and forth over this - I think it was me that started the breakage. I realised this at ~11pm last night after some hours of picking through and - boy, oh boy - that wasn't a fun realisation. Disclosure follows for completeness.

The short version: it appears I inadvertently pulled from `dev`, didn't switch the base branch to `4.8.2`, and committed to `4.8.2`, which then took all the post-`4.8.2`-creation date commits on `dev` along with it. There is no `dev`-to-`4.8.2` merging. Rather, there wasn't until I did it yesterday. Where merges are done, they go the other way - `4.8.2`-to-`dev`. Ugh. Hence the references to `4.9` and other 'future' stuff that shouldn't even have been in `4.8.2`. If you trace the network graph yesterday, you can see where it happened. I made the changes to include release dates in history (and the irony of me asking where I should do this _to explicitly avoid fecking it all up_ is not lost on me).

Yeah. So, likely my fault that this reared its head, though I can't say what the cherry picking did, I got into a muddle near midnight last night and decided to sleep on it…and @Bloke fixed it overnight. I owe you beers, Stef. Thank you.

Sorry for breaking it. Won't be doing that again!

**Comment by bloatware** [comment]
@Bloke @petecooper thanks for your Hercules work - the man has triumphed over the machine!

**Comment by petecooper** [comment]
Also, `dev-demo` site is now running the fixed 4.8.2 branch.

**Comment by Bloke** [comment]
No worries, don't let it put you off committing. It's just one of those things.

The annoying part is that it _should_ have been reversible with a revert commit of some kind. Indeed, I was hoping it was still unpickable by being cleverer and outsmarting git, but it had other ideas and wanted to be cleverer-er.

If the blaze had been confined to the 4.8.2 branch, it might have been okay to resync and pretend it hadn't happened. But I suspect as soon as the branch was merged/picked into dev, it brought all the anti-dev payload with it, and then git wasn't able to distinguish between which was the most recent branch, as they were largely the same: 4.8.2. The fact that 4.8.2 didn't have any 4.9 features in it effectively undid all our 4.9 work to date on the dev branch, eeek!

Anyway. If anyone would like to sift through [this comment](https://github.com/textpattern/textpattern/issues/1516#issuecomment-657111081) be my guest. That represents every commit in the new `482-recovery-bloke` branch. Please flag if there's anything obvious missing and I'll see to it. Or, y'know, just test the hell out of it on the new dev-demo (thanks, Pete).

**Comment by petecooper** [comment]
>No worries, don't let it put you off committing. It's just one of those things.

When I buy you that beer, I tell you about how I took a global CDN offline with a rogue commit that didn't fire a JSON linter and…well, you live and learn!

**Comment by Bloke** [comment]
Ouch!

**Comment by philwareham** [comment]
Don’t worry too much. We’ve all broken code repos at some point in the past. It can be quite fragile. Thanks to all for fixing.

**Comment by petecooper** [comment]
Looking at the outstanding issue queue for 4.8.2, there's currently one thing: https://github.com/textpattern/textpattern/issues/1510

Is https://github.com/textpattern/textpattern/commit/7a3989e8096314d8afe5157e2cf68e76ecc9ff0b enough of a short-term fix, and if so can the issue be retargeted at 4.9 or beyond?

**Comment by Bloke** [comment]
Yep, I'm fine with the belt and braces approach for this now. It satisfies the OP and gets rid of a long-standing bug.

We'll revisit it in 4.9 and introduce a proper class that we can use across the board for fetching update info, plugins from the repo, perhaps themes from its repo too, and anywhere else we're allowing URLs to be treated as content.

**Comment by petecooper** [comment]
Thanks @Bloke. In that case, the issue queue is clear for 4.8.2. There are a few items on the docs to-do list, but they're not blockers.

**Comment by philwareham** [comment]
Is there any new translations strings needed for 4.8.2 over and above the additions a couple of weeks ago? If not then I will tag the Textpack/Pophelp repos with a release tag (although it makes no difference to the main repo if I tag or not).

**Comment by philwareham** [comment]
Textpack/Pophelp now tagged with 4.8.2 in readiness.

**Comment by petecooper** [comment]
Adding https://github.com/textpattern/textpattern/issues/1518 as a blocker.

**Comment by bloatware** [comment]
#1518 is fixed in my tests, but extra eyes are welcome.

**Comment by Bloke** [comment]
#1518 appears to be fixed through cursory testing. Please stress test multi-edit and shift-select checkboxes on all panels, even when searching, saving, paging, etc.

**Comment by petecooper** [comment]
Any benefit in pushing the demo site update frequency to hourly in the final few furlongs before release?

**Comment by Bloke** [comment]
> Any benefit in pushing the dev demo site update frequency to hourly in the final few furlongs before release?

Hmmm, maybe, maybe not. It gives people less time before a reset to play. I'd leave it as-is but if you wanna tinker, by all means.

**Comment by petecooper** [comment]
I'll leave it as-is. I was thinking the same, but wanted to gauge the room, as it were.

**Comment by Bloke** [comment]
We also need to stress test 9dfc203c and af6a15cee using as many tags and attributes as we can. I tried many versions of `break` with and without values, and with weird (and some invalid) constructs yesterday and it behaved as expected. I encourage others to do the same.

Haven't had a chance to try out the `yield` thing yet but I'm working on a site today that uses `yield` and `if_yield` (but not yield-else) so I'll see if anything crops up. Anyone who can grill these tags and output_form/short tags in general please do so.

**Comment by Bloke** [comment]
@bloatware Can you please give me a quick rundown of what the Detect if_yield else (ac509993) and Waiting for unique (ffca35b9) commits bring to the table? So I can document them in History.txt. And some examples of them in use in tags, so I can build on those tests.

Looks as if `getIfElse()` caches results for better performance? Any other benefits to us/plugin authors for using that function over `parse()`?

Thanks.

**Comment by bloatware** [comment]
@Bloke `getIfElse()` is just getting `true/false` parts of `thing`, without processing them. Unlike deprecated `EvalElse()`, it is not intended to be used in pre-4.6 `parse(EvalElse($thing, $condition))` construction, replaced with `parse($thing, $condition)`, because this wouldn't properly handle global attributes. Its main purpose atm is to check whether `else` part is set in `<txp:if_yield else />`, but other tags could use it to this aim too.
```
getIfElse('<txp:tag>hello</txp:tag>') returns null
getIfElse('<txp:tag>hello<txp:else /></txp:tag>') returns ''
```
As for `unique`, I have abused of valueless global `replace` as alias of `array_filter()` and `do_list_unique()`:
```
<txp:variable name="test" value="2,1,0,2,1" replace output /> outputs 2,1,1,2
<txp:variable name="test" value="2,1,0,2,1" trim replace output /> outputs 2,1,0
```
Or should we switch them? The second one seems more useful in txp.

**Comment by Bloke** [comment]
@bloatware Thank you. `getIfElse()` is going to be handy for smd_xml and/or smd_query I expect. I remember having to do some `EvalElse()` hoop-jumping to get the else portion in at least one of those plugins.

Nice hack for `replace`! I'm not sure which is best. Probably the way you have it, as 'trim' will remove zeros and empties as well, right?

**Comment by petecooper** [comment]
*Go* or *No Go* for 4.8.2?

@bloatware @Bloke @philwareham - please mark this post with a thumbs up for *Go* or thumbs down for *No Go*.

No urgency. Thank you.

**Comment by Bloke** [comment]
I went with 'eyes' instead of thumbs up/down. Still looking, but I'm leaning closer to thumbs up than down.

The more people testing this (especially tags) the better. Incidentally, Pete, please check issue #671  comments.

**Comment by petecooper** [comment]
NTS: We should sort a pre-release testing doc for interested parties before next non-patch release. Not unit testing as such, use enthusiast testers etc.

**Comment by bloatware** [comment]
Not yet, sorry. I really don't like https://github.com/textpattern/textpattern/commit/6352b00c8de87ec1dac78dcf6c40e6f4a6651492, but see no other fix atm.

**Comment by bloatware** [comment]
Okay, it should be rather safe now. Should we do a last 'speak now' pre-release forum call for bugs/features?

**Comment by petecooper** [comment]
Good idea, @bloatware - OK for me to handle that, or would you like to handle it?

**Comment by petecooper** [comment]
I'll refresh the .com article tonight to take into account recent changes.

**Comment by petecooper** [comment]
_Go_ or _No Go_ for 4.8.2?

@bloatware @Bloke @philwareham - please mark this post with a thumbs up for _Go_ or thumbs down for _No Go_.

No urgency. Thank you very much.

**Comment by bloatware** [comment]
I would like to explore `searchall` attribute a little bit, let us put the deadline on this week-end.

**Comment by bloatware** [comment]
Final testing this evening and it will be Go on my side.

**Comment by petecooper** [comment]
We're so close now…can we cross the line?

### _Go_ or _No Go_ for 4.8.2?

@bloatware @Bloke @philwareham - please mark this post with a thumbs up for _Go_ or thumbs down for _No Go_.

Important but not urgent. Thank you very much.

(This supersedes all previous _Go_ or _No Go_ posts from this thread.)

**Comment by philwareham** [comment]
@petecooper oh, while I remember - can you push the latest server config live. I've relaxed the CSP again for scripts and styles so the .com site admin panel works as expected again. Cheers.

**Comment by petecooper** [comment]
FYI - we're all _Go_ on https://github.com/textpattern/textpattern/issues/1513#issuecomment-662905969

**Comment by philwareham** [comment]
Any news on a release plan? Ta!

**Comment by petecooper** [comment]
I'm currently updating the announcement article with latest history stuff, I'll defer to @Bloke for the button pushing sit rep.

**Comment by Bloke** [comment]
I fixed another niggle Adi raised after the one he mentioned over the weekend. Part of his request will have to wait (#1526).

As far as I'm concerned, it's all good to go.

**Comment by philwareham** [comment]
OK, can you please run the release process when you get a few minutes?

**Comment by petecooper** [comment]
FYI added item to OP to include latest pophelp.

**Comment by philwareham** [comment]
OK, pophelps updated now. Thanks.

**Comment by petecooper** [comment]
No further pophelp amends from me.

**Comment by Bloke** [comment]
Oki doke. Thanks for those pophelp changes.

I've given the release doc on .com a sweep and amended a few minor things. Thanks for writing that, Pete, it's ace.

Let's freeze commits to 482-recovery-bloke as of now then. I'll bundle up a release this afternoon and run some final tests then hit the button, all being well, so the 4.8.2 train can roll into the station.

P.S. @bloatware If you get a little time would you please have a go at merging dev into the custom-fields branch. I had a play the other day but there was some stuff (in publish.php, iirc) that I couldn't untangle and feared I'd screw up as it was late at night. Thanks.

**Comment by bloatware** [comment]
@Bloke done, hope well.

**Comment by Bloke** [comment]
Awesome, thanks. I'm sure it's fine :)

**Comment by Bloke** [comment]
@bloatware This is probably not the right place to post this but I'm just trying out the category match changes. Can I just confirm something (default site setup). If I add:

``` html
<txp:article_custom match="Category" category='<txp:category />' />
```

to the default template and visit these in turn:
```
[1] /category/hope-for-the-future
[2] /category/meaningful-labor
[3] /category/everything-is-going-to-be-alright
```
Then [3] fails, and [1] and [2] return the article, as expected. It's matching _either_ category1 OR category2. Same result if I remove the `match` attribute from the tag.

With this:

``` html
<txp:article_custom match="Category1" category='<txp:category />' />
```

Then [2] and [3] fail (as expected) and [1] passes (as expected).

With this:

``` html
<txp:article_custom match="Category2" category='<txp:category />' />
```

Then [1] and [3] fail (as expected) and [2] passes (as expected).

With this:

``` html
<txp:article_custom match="Category1,Category2" category='<txp:category />' />
```

Then they all fail because it's matching BOTH cat1 and cat2. So how would I specify this on the URL so that the test passes? Do I need to use a specific permlink scheme (e.g. breadcrumb?) for the default page? No matter what I've tried so far, I can't get anything to match. What am I missing?

Thanks!

**Comment by bloatware** [comment]
@Bloke there are two ways to achieve it:
``` txp
<txp:article_custom match="Category1,Category2" category="hope-for-the-future, meaningful-labor" />
```
should output the articles with both Category1 and Category2 in `(hope-for-the-future, meaningful-labor)`.

If you need to pull it from URL, you should indicate what parts of it must match Category1 and Category2:
``` txp
<txp:article_custom match="Category1=c1,Category2=c2" />
```
should output what expected on `?c1=hope-for-the-future&c2=meaningful-labor` page. I also think that
``` txp
<txp:article_custom match="Category1=2,Category2=3" />
```
should work on `category/hope-for-the-future/meaningful-labor/` page.

**Comment by Bloke** [comment]
Ah okay, thanks. Knew there had to be a way to do it. Not really matching expectations though in my tests. With:

``` html
<txp:article_custom match="Category1=2,Category2=3" />
```

And these tests as input:
```
[1] /category/hope-for-the-future
[2] /category/meaningful-labor
[3] /category/everything-is-going-to-be-alright
[4] /category/hope-for-the-future/meaningful-labor
[5] /category/meaningful-labor/hope-for-the-future
```
I get:
[1] passed
[2] nothing (expected)
[3] nothing (expected)
[4] passed (expected)
[5] nothing (expected)

I was expecting 1 to fail given this is an AND situation? Is it?

With:
``` html
<txp:article_custom match="Category1=c1,Category2=c2" />
```

And these tests as input:
```
[1] /category/hope-for-the-future
[2] /category/meaningful-labor
[3] /category/everything-is-going-to-be-alright
[4] /category/hope-for-the-future/meaningful-labor
[5] /category/meaningful-labor/hope-for-the-future
[6] /?c1=hope-for-the-future
[7] /?c2=hope-for-the-future
[8] /?c1=meaningful-labor
[9] /?c2=meaningful-labor
[10] /?c1=hope-for-the-future&c2=meaningful-labor
[11] /?c1=meaningful-labor&c2=hope-for-the-future
[12] /?c1=everything-is-going-to-be-alright
[13] /?c1=hope-for-the-future&c2=everything-is-going-to-be-alright
[14] /?c1=hope-for-the-future&c2=meaningful-labor&c3=everything-is-going-to-be-alright
```
I get:
[1] passed
[2] passed
[3] passed
[4] passed
[5] passed
[6] passed
[7] nothing (expected)
[8] nothing (expected)
[9] passed
[10] passed (expected)
[11] nothing (expected)
[12] nothing (expected)
[13] nothing (expected)
[14] passed (expected)

Seems to be doing an OR still. Is that expected? Does it only match AND if you _also_ include a `category` attribute? When I'm using that and explicitly matching using a category, then ALL these pass, as expected, since there's no implied order:

``` html
<txp:article_custom match="Category1,Category2" category="hope-for-the-future, meaningful-labor" />
<txp:article_custom match="Category1,Category2" category="meaningful-labor, hope-for-the-future" />
<txp:article_custom match="Category1,Category2" category="everything-is-going-to-be-alright, hope-for-the-future, meaningful-labor" />
```

**Comment by Bloke** [comment]
@bloatware For clarity, I'm not super fussed if this ORish behaviour is expected and goes into 4.8.2. Just want to confirm that this is intended behaviour so it can be documented in the tag docs.

**Comment by bloatware** [comment]
@Bloke when a parameter (say `c2`) is not in URL, there is no filtering. Try `/?c1=hope-for-the-future&c2=` instead.. nope, working only for cf, not categories. Should we fix it?

**Comment by Bloke** [comment]
That's what I mean by 'working'. If there's something not present then it can't very well filter by it. But if we're saying that AND is the behaviour when you use two items in `match`, and it works for hard-coded `category` names then, well, maybe it should be consistent.

Then again, if it works by supplying an empty parameter on the URL, then I see that as an acceptable trade-off. I mean, you've made a conscious decision to state that you're matching both so I suppose you should be expected to supply both on the URL if that's your use case. RIght? Otherwise, it will only match one OR the other of those supplied. Tat makes sense logically doesn't it?

But if it's easy to change it, then feel free.

**Comment by bloatware** [comment]
We can change it either way, but then the same rules should apply to cf. So `<txp:article match="cf1,cf2" />` on `?cf1=a` page should return the articles with `cf1=a` and 'empty' `cf2`. But is `0` 'empty' here? And how would we retrieve *all* articles with `cf1=a`? Via `?cf1=a&cf2=%`? But what if `cf2` is an array?

**Comment by Bloke** [comment]
Yeah. A minefield. Like you say, what is 'empty'? I'd be tempted to leave it as the most bwc version we can, and just fall back on matching OR if there's missing elements. I don't see there's much else we can do. I'm open to suggestions on 'best' behaviour.

**Comment by bloatware** [comment]
Re custom fields, I'm more or less convinced that a search for `?colour=red` should return all 'red' articles, whatever their size, price etc. But is it also true for categories ? Dunno..

**Comment by Bloke** [comment]
I'm inclined to agree. It should be the same for cats if you only specify some of the information to match. But if you specify two or more items, it has to match ALL of what you've given if you've asked it to match multiple items. So if you supply two out of three, then it pulls things that match the two you've supplied based on which cats you've assigned to be matched. If you supply all three, it has to match them all. Does that make sense? Kind of 'progressive AND'?

Does that trample on OR situations?

**Comment by bloatware** [comment]
That's right, it matches all supplied filters. OR was a legacy inter-categories exception, other criteria were always connected by AND. Presently, if `match` contains `Category`, we get
```
(Category1 IN(...) OR Category2 IN(...)) AND ... AND ...
```
but otherwise
```
(Category1 IN(...) AND Category2 IN(...)) AND ... AND ...
```

**Comment by Bloke** [comment]
All good thanks. Let's leave it as-is for cats. It makes sense.

On the custom fields front, matching using `match="custom1=cf1,custom2=cf2"` works a treat in exactly the same way. One more thing to verify. Are we expecting to be able to use CFs via positional URL params? e.g. if I set up a 'search' article to act as a search endpoint inside a 'filter' section: example.org/filter/search.

Could I do something like this?

```
<txp:article_custom match="custom1=3,custom2=4" wraptag="ul" break="li">
<txp:title />
</txp:article_custom>
```

And then use these URLs to filter out articles:
```
[1] /filter/search/Blue
[2] /filter/search/Red
[3] /filter/search/Blue/Jeans
...
```

i.e. matching custom1 as a colour and custom2 as a thing in a fictitious product site.

At the moment all I get is 404 city. And if I try and do it on the landing page:
```
[1] /filter/Blue/
[2] /filter/Blue/Jeans/
```

I get 404 too, presumably because it's looking for article title 'Blue'? So are custom fields only allowed to match via traditional URL question mark params?

**Comment by bloatware** [comment]
CF can match path chunks, but yes, txp is probably searching for 'Blue' article in this case. If such an article were existing, you should be able to match `Jeans` against `custom2` on its page. Tough-tough..

**Comment by Bloke** [comment]
Yeah, I'm just pushing boundaries now! Let's leave it. This is working well in my tests. Are you happy with it?

**Comment by bloatware** [comment]
That's strange, positional params work for me.. once I manage to land on a right place.

I don't see how txp could guess what `/filter/Blue/Jeans/` means. If `Blue` article does not exist, we could land on `filter` section page instead of 404, but I'm not sure it's ok bwc- and seo-wise. Currently all permlink modes recognise `/section/title` pattern.

Trailing slash = article list rule would help, but again, old `/section` and `.../title/` bookmarks would become invalid.

So yep, rewrite rules to rescue and let us be happy with it.

**Comment by Bloke** [comment]
Cool. I'm probably not being clever enough. If you're happy it works I'll get this bundled up and released.

**Comment by Bloke** [comment]
Oki doke, 4.8.2 released. Thank you as ever for all your hard work and diligence at making this version rock. Another fab team effort. Drinks are on me.

**Comment by petecooper** [comment]
OP is done. Closing!
